torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,117 @@
+import contextlib
+from typing import Dict, Optional, Tuple
+
+import torch
+import torch.distributed as dist
+import torch.distributed._functional_collectives as ft_c
+from torch.distributed._tensor import DTensor
+from torch.distributed.device_mesh import DeviceMesh
+
+aten = torch.ops.aten
+
+
+def sdpa_handler(
+    op_call: torch._ops.OpOverload,
+    args: Tuple[object, ...],
+    kwargs: Dict[str, object],
+) -> object:
+    # extract local tensor and sharding infos to a OpInfo
+    op_info = DTensor._op_dispatcher.unwrap_to_op_info(op_call, args, kwargs)
+
+    # sharding propagation
+    DTensor._op_dispatcher.sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
```

Comment:
I think the reason why you don't have collective is that you only run sharding propagation, but the actual resharding not happened yet. because we are in custom handler, you would need to add the resharding logic to the custom handler too https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/dispatch.py#L165-L178

Reply:
iiuc we still need to write the redistribute logic so that it won't fail for the cases where the input shardings are not expected?

Reply:
Right now we need to support Replicate() and Shard(2) due to issues with view and nn.Linear for nn.MultiHeadedAttention. If we don't care about that we can correctly only support Shard(2)

Reply:
Sounds good let's follow up later